File: python_expense_tracker/tests/test_functional.py
"""
Functional tests for Expense Tracking System
Tests user stories and end-to-end functional scenarios
"""

import pytest

import models
import expense_operations
import utils


def test_user_can_record_expense(clean_db):
    """
    User Story: As a user, I want to record an expense
    so that I can track my spending
    """
    # Given: A user wants to record an expense
    user_name = "Alice"
    category = "Food"
    title = "Lunch at restaurant"
    amount = 45.50
    date = "2025-10-25"
    
    # When: The user records the expense
    expense_id = expense_operations.record_expense(
        date=date,
        category=category,
        title=title,
        amount=amount,
        user_name=user_name
    )
    
    # Then: The expense should be successfully recorded
    assert expense_id > 0, "Expense should be recorded with valid ID"
    
    # And: The expense should be retrievable
    expenses = expense_operations.view_all_expenses()
    assert len(expenses) == 1, "Should have one expense"
    
    recorded_expense = expenses[0]
    assert recorded_expense['user_name'] == user_name
    assert recorded_expense['category_name'] == category
    assert recorded_expense['title'] == title
    assert recorded_expense['amount'] == amount
    assert recorded_expense['date'] == date


def test_user_can_view_expenses_by_date(clean_db):
    """
    User Story: As a user, I want to view expenses within a date range
    so that I can see my spending for a specific period
    """
    # Given: Multiple expenses exist across different dates
    user_id = models.create_user("Bob")
    categories = models.get_all_categories()
    category_id = categories[0]['id']
    
    models.insert_expense('2025-10-20', category_id, 'Expense 1', 100.00, user_id)
    models.insert_expense('2025-10-22', category_id, 'Expense 2', 50.00, user_id)
    models.insert_expense('2025-10-24', category_id, 'Expense 3', 75.00, user_id)
    models.insert_expense('2025-10-26', category_id, 'Expense 4', 25.00, user_id)
    
    # When: The user filters expenses by date range
    filtered_expenses = expense_operations.view_expenses_by_date(
        min_date='2025-10-21',
        max_date='2025-10-25'
    )
    
    # Then: Only expenses within the date range should be returned
    assert len(filtered_expenses) == 2, "Should return 2 expenses in date range"
    
    for expense in filtered_expenses:
        assert '2025-10-21' <= expense['date'] <= '2025-10-25', \
            "All expenses should be within date range"


def test_user_can_view_summary(clean_db):
    """
    User Story: As a user, I want to view a summary of my expenses
    so that I can understand my spending patterns
    """
    # Given: Multiple users have recorded expenses in different categories
    user1_id = models.create_user("Charlie")
    user2_id = models.create_user("Diana")
    
    categories = models.get_all_categories()
    food_cat = categories[0]
    transport_cat = categories[1]
    
    # Charlie's expenses
    models.insert_expense('2025-10-25', food_cat['id'], 'Groceries', 100.00, user1_id)
    models.insert_expense('2025-10-25', transport_cat['id'], 'Gas', 50.00, user1_id)
    
    # Diana's expenses
    models.insert_expense('2025-10-25', food_cat['id'], 'Restaurant', 75.00, user2_id)
    models.insert_expense('2025-10-25', transport_cat['id'], 'Uber', 25.00, user2_id)
    
    # When: The user requests an expense summary
    summary = expense_operations.calculate_summary()
    
    # Then: The summary should show correct totals
    assert summary['total'] == 250.00, "Total should be sum of all expenses"
    assert summary['count'] == 4, "Should have 4 expenses"
    
    # And: The summary should break down by user
    assert summary['by_user']['Charlie'] == 150.00, "Charlie's total should be 150.00"
    assert summary['by_user']['Diana'] == 100.00, "Diana's total should be 100.00"
    
    # And: The summary should break down by category
    assert summary['by_category'][food_cat['name']] == 175.00, "Food total should be 175.00"
    assert summary['by_category'][transport_cat['name']] == 75.00, "Transport total should be 75.00"
    
    # And: The summary should group expenses by user
    assert 'user_expenses' in summary
    assert len(summary['user_expenses']['Charlie']) == 2, "Charlie should have 2 expenses"
    assert len(summary['user_expenses']['Diana']) == 2, "Diana should have 2 expenses"


def test_user_can_manage_categories(clean_db):
    """
    User Story: As a user, I want to manage categories
    so that I can organize my expenses
    """
    # Given: Default categories exist
    initial_categories = models.get_all_categories()
    initial_count = len(initial_categories)
    assert initial_count >= 7, "Should have default categories"
    
    # When: The user adds a new category
    new_category_id = models.create_category("Custom Category")
    
    # Then: The category should be created
    assert new_category_id > 0, "Category should be created with valid ID"
    
    # And: The category should appear in the list
    all_categories = models.get_all_categories()
    assert len(all_categories) == initial_count + 1, "Should have one more category"
    
    category_names = [cat['name'] for cat in all_categories]
    assert "Custom Category" in category_names, "New category should be in list"
    
    # And: The user can retrieve the category by name
    retrieved_category = models.get_category_by_name("Custom Category")
    assert retrieved_category is not None, "Should find the category"
    assert retrieved_category['name'] == "Custom Category"


def test_user_can_manage_users(clean_db):
    """
    User Story: As a user, I want to manage users
    so that multiple people can track their expenses
    """
    # Given: No users exist initially
    initial_users = models.get_all_users()
    assert len(initial_users) == 0, "Should start with no users"
    
    # When: Multiple users are added
    user1_id = models.create_user("Emma")
    user2_id = models.create_user("Frank")
    user3_id = models.create_user("Grace")
    
    # Then: All users should be created
    assert user1_id > 0 and user2_id > 0 and user3_id > 0, "All users should have valid IDs"
    
    # And: All users should appear in the list
    all_users = models.get_all_users()
    assert len(all_users) == 3, "Should have 3 users"
    
    user_names = [user['name'] for user in all_users]
    assert "Emma" in user_names, "Emma should be in list"
    assert "Frank" in user_names, "Frank should be in list"
    assert "Grace" in user_names, "Grace should be in list"
    
    # And: Users can be retrieved by name
    emma = models.get_user_by_name("Emma")
    assert emma is not None, "Should find Emma"
    assert emma['name'] == "Emma"


def test_expense_validation_prevents_invalid_data(clean_db):
    """
    User Story: As a system, I want to validate expense data
    so that only valid expenses are recorded
    """
    # Given: A user attempts to record expenses with invalid data
    user_id = models.create_user("Validator")
    categories = models.get_all_categories()
    category_id = categories[0]['id']
    
    # When/Then: Invalid date format should be rejected
    with pytest.raises(ValueError):
        expense_operations.record_expense(
            date='invalid-date',
            category=categories[0]['name'],
            title='Test',
            amount=50.00,
            user_name='Validator'
        )
    
    # When/Then: Negative amount should be rejected
    with pytest.raises(ValueError):
        expense_operations.record_expense(
            date='2025-10-25',
            category=categories[0]['name'],
            title='Test',
            amount=-10.00,
            user_name='Validator'
        )
    
    # When/Then: Zero amount should be rejected
    with pytest.raises(ValueError):
        expense_operations.record_expense(
            date='2025-10-25',
            category=categories[0]['name'],
            title='Test',
            amount=0.00,
            user_name='Validator'
        )
    
    # When/Then: Empty title should be rejected
    with pytest.raises(ValueError):
        expense_operations.record_expense(
            date='2025-10-25',
            category=categories[0]['name'],
            title='',
            amount=50.00,
            user_name='Validator'
        )
    
    # When/Then: Empty user name should be rejected
    with pytest.raises(ValueError):
        expense_operations.record_expense(
            date='2025-10-25',
            category=categories[0]['name'],
            title='Test',
            amount=50.00,
            user_name=''
        )
    
    # And: No invalid expenses should be recorded
    expenses = expense_operations.view_all_expenses()
    assert len(expenses) == 0, "No invalid expenses should be recorded"


def test_user_can_filter_by_amount_range(clean_db):
    """
    User Story: As a user, I want to filter expenses by amount
    so that I can find large or small expenses
    """
    # Given: Multiple expenses with different amounts exist
    user_id = models.create_user("Henry")
    categories = models.get_all_categories()
    category_id = categories[0]['id']
    
    models.insert_expense('2025-10-25', category_id, 'Small 1', 5.00, user_id)
    models.insert_expense('2025-10-25', category_id, 'Medium 1', 50.00, user_id)
    models.insert_expense('2025-10-25', category_id, 'Large 1', 500.00, user_id)
    models.insert_expense('2025-10-25', category_id, 'Medium 2', 75.00, user_id)
    models.insert_expense('2025-10-25', category_id, 'Small 2', 10.00, user_id)
    
    # When: The user filters by amount range
    medium_expenses = expense_operations.view_expenses_by_amount(
        min_amount=40.00,
        max_amount=100.00
    )
    
    # Then: Only expenses within the amount range should be returned
    assert len(medium_expenses) == 2, "Should return 2 medium expenses"
    
    for expense in medium_expenses:
        assert 40.00 <= expense['amount'] <= 100.00, \
            "All expenses should be within amount range"


def test_user_can_filter_by_category(clean_db):
    """
    User Story: As a user, I want to filter expenses by category
    so that I can see spending in specific areas
    """
    # Given: Expenses exist in multiple categories
    user_id = models.create_user("Iris")
    categories = models.get_all_categories()
    
    cat1 = categories[0]
    cat2 = categories[1]
    cat3 = categories[2]
    
    models.insert_expense('2025-10-25', cat1['id'], 'Cat1 Exp1', 100.00, user_id)
    models.insert_expense('2025-10-25', cat2['id'], 'Cat2 Exp1', 50.00, user_id)
    models.insert_expense('2025-10-25', cat1['id'], 'Cat1 Exp2', 75.00, user_id)
    models.insert_expense('2025-10-25', cat3['id'], 'Cat3 Exp1', 25.00, user_id)
    models.insert_expense('2025-10-25', cat2['id'], 'Cat2 Exp2', 60.00, user_id)
    
    # When: The user filters by a specific category
    cat1_expenses = expense_operations.view_expenses_by_category([cat1['name']])
    
    # Then: Only expenses from that category should be returned
    assert len(cat1_expenses) == 2, "Should return 2 expenses from category 1"
    
    for expense in cat1_expenses:
        assert expense['category_name'] == cat1['name'], \
            "All expenses should be from the specified category"
    
    # When: The user filters by multiple categories
    multi_cat_expenses = expense_operations.view_expenses_by_category([cat1['name'], cat2['name']])
    
    # Then: Expenses from all specified categories should be returned
    assert len(multi_cat_expenses) == 4, "Should return 4 expenses from both categories"


def test_user_can_track_expenses_over_time(clean_db):
    """
    User Story: As a user, I want to track expenses over time
    so that I can see spending trends
    """
    # Given: Expenses are recorded over multiple months
    user_id = models.create_user("Jack")
    categories = models.get_all_categories()
    category_id = categories[0]['id']
    
    # October expenses
    models.insert_expense('2025-10-05', category_id, 'Oct Week 1', 100.00, user_id)
    models.insert_expense('2025-10-12', category_id, 'Oct Week 2', 150.00, user_id)
    models.insert_expense('2025-10-19', category_id, 'Oct Week 3', 200.00, user_id)
    models.insert_expense('2025-10-26', category_id, 'Oct Week 4', 175.00, user_id)
    
    # November expenses
    models.insert_expense('2025-11-02', category_id, 'Nov Week 1', 125.00, user_id)
    models.insert_expense('2025-11-09', category_id, 'Nov Week 2', 180.00, user_id)
    
    # When: The user views expenses for October
    october_expenses = expense_operations.view_expenses_by_date(
        min_date='2025-10-01',
        max_date='2025-10-31'
    )
    
    # Then: Only October expenses should be returned
    assert len(october_expenses) == 4, "Should have 4 October expenses"
    
    october_summary = expense_operations.calculate_summary(october_expenses)
    assert october_summary['total'] == 625.00, "October total should be 625.00"
    
    # When: The user views expenses for November
    november_expenses = expense_operations.view_expenses_by_date(
        min_date='2025-11-01',
        max_date='2025-11-30'
    )
    
    # Then: Only November expenses should be returned
    assert len(november_expenses) == 2, "Should have 2 November expenses"
    
    november_summary = expense_operations.calculate_summary(november_expenses)
    assert november_summary['total'] == 305.00, "November total should be 305.00"


def test_multiple_users_independent_tracking(clean_db):
    """
    User Story: As a user, I want my expenses tracked separately from other users
    so that each person's spending is independent
    """
    # Given: Multiple users record expenses
    user1_id = models.create_user("Kate")
    user2_id = models.create_user("Leo")
    
    categories = models.get_all_categories()
    category_id = categories[0]['id']
    
    # Kate's expenses
    models.insert_expense('2025-10-25', category_id, 'Kate Exp 1', 100.00, user1_id)
    models.insert_expense('2025-10-25', category_id, 'Kate Exp 2', 50.00, user1_id)
    
    # Leo's expenses
    models.insert_expense('2025-10-25', category_id, 'Leo Exp 1', 200.00, user2_id)
    models.insert_expense('2025-10-25', category_id, 'Leo Exp 2', 75.00, user2_id)
    models.insert_expense('2025-10-25', category_id, 'Leo Exp 3', 25.00, user2_id)
    
    # When: Each user views their expenses
    kate_expenses = expense_operations.view_expenses_by_user(user1_id)
    leo_expenses = expense_operations.view_expenses_by_user(user2_id)
    
    # Then: Each user should only see their own expenses
    assert len(kate_expenses) == 2, "Kate should have 2 expenses"
    assert len(leo_expenses) == 3, "Leo should have 3 expenses"
    
    # And: Expenses should belong to the correct user
    for expense in kate_expenses:
        assert expense['user_name'] == 'Kate', "All expenses should belong to Kate"
    
    for expense in leo_expenses:
        assert expense['user_name'] == 'Leo', "All expenses should belong to Leo"
    
    # And: Totals should be calculated correctly per user
    summary = expense_operations.calculate_summary()
    assert summary['by_user']['Kate'] == 150.00, "Kate's total should be 150.00"
    assert summary['by_user']['Leo'] == 300.00, "Leo's total should be 300.00"

File: cpp_expense_tracker/tests/test_functional.cpp
/**
 * test_functional.cpp
 * Functional tests based on user stories for C++ Expense Tracker
 * 
 * C++ TESTING FEATURES DEMONSTRATED:
 * - User story-based testing
 * - Complete application flow testing
 * - Real-world scenario testing
 * - End-user perspective testing
 */

#include <gtest/gtest.h>
#include "test_helpers.h"
#include "../expense_operations.h"
#include "../utils.h"

// ============================================================================
// USER STORY TESTS
// ============================================================================

/**
 * USER STORY: As a user, I can record an expense
 * Given: I have expense details (date, category, title, amount)
 * When: I record the expense
 * Then: The expense is saved and I can view it
 * 
 * C++ FEATURE: Complete workflow with RAII and STL
 */
TEST_F(ModelsTest, UserCanRecordExpense) {
    // User records an expense
    int expense_id = ExpenseOperations::recordExpense(
        *db, "2025-10-25", "Groceries", "Weekly Shopping", 150.0, "JohnDoe"
    );
    
    // Expense is saved
    ASSERT_GT(expense_id, 0);
    
    // User can view the expense
    std::vector<Expense> expenses = ExpenseOperations::viewAllExpenses(*db);
    ASSERT_EQ(expenses.size(), 1);
    
    // Verify expense details
    EXPECT_EQ(expenses[0].date, "2025-10-25");
    EXPECT_EQ(expenses[0].title, "Weekly Shopping");
    EXPECT_DOUBLE_EQ(expenses[0].amount, 150.0);
    EXPECT_EQ(expenses[0].user_name, "JohnDoe");
    EXPECT_EQ(expenses[0].category_name, "Groceries");
}

/**
 * USER STORY: As a user, I can view expenses by date range
 * Given: I have multiple expenses on different dates
 * When: I filter by date range
 * Then: I see only expenses within that range
 */
TEST_F(ModelsTest, UserCanViewExpensesByDate) {
    // User records expenses on different dates
    ExpenseOperations::recordExpense(*db, "2025-10-20", "Food", "Breakfast", 10.0, "User");
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "Lunch", 20.0, "User");
    ExpenseOperations::recordExpense(*db, "2025-10-30", "Food", "Dinner", 30.0, "User");
    
    // User filters by date range (Oct 22-28)
    std::vector<Expense> filtered = ExpenseOperations::viewExpensesByDate(
        *db, "2025-10-22", "2025-10-28"
    );
    
    // User sees only expenses in range
    ASSERT_EQ(filtered.size(), 1);
    EXPECT_EQ(filtered[0].title, "Lunch");
    EXPECT_EQ(filtered[0].date, "2025-10-25");
}

/**
 * USER STORY: As a user, I can view expense summary
 * Given: I have recorded multiple expenses
 * When: I view the summary
 * Then: I see total amount, count, and breakdown by category and user
 */
TEST_F(ModelsTest, UserCanViewSummary) {
    // User records multiple expenses
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "Lunch", 25.0, "Alice");
    ExpenseOperations::recordExpense(*db, "2025-10-26", "Transport", "Taxi", 15.0, "Bob");
    ExpenseOperations::recordExpense(*db, "2025-10-27", "Food", "Dinner", 40.0, "Alice");
    
    // User views summary
    std::vector<Expense> expenses = ExpenseOperations::viewAllExpenses(*db);
    ExpenseSummary summary = ExpenseOperations::calculateSummary(*db, &expenses);
    
    // User sees total and count
    EXPECT_DOUBLE_EQ(summary.total, 80.0);
    EXPECT_EQ(summary.count, 3);
    
    // User sees breakdown by user
    EXPECT_EQ(summary.by_user.size(), 2);
    EXPECT_DOUBLE_EQ(summary.by_user["Alice"], 65.0);
    EXPECT_DOUBLE_EQ(summary.by_user["Bob"], 15.0);
    
    // User sees breakdown by category
    EXPECT_DOUBLE_EQ(summary.by_category["Food"], 65.0);
    EXPECT_DOUBLE_EQ(summary.by_category["Transport"], 15.0);
    
    // User sees individual expenses per user
    EXPECT_EQ(summary.user_expenses["Alice"].size(), 2);
    EXPECT_EQ(summary.user_expenses["Bob"].size(), 1);
}

/**
 * USER STORY: As a user, I can manage categories
 * Given: The system has default categories
 * When: I create a new category
 * Then: I can use it for expenses
 */
TEST_F(ModelsTest, UserCanManageCategories) {
    // System has default categories
    std::vector<Category> categories = Models::getAllCategories(*db);
    size_t initial_count = categories.size();
    EXPECT_GT(initial_count, 0);
    
    // User creates new category
    int new_category_id = Models::createCategory(*db, "CustomCategory");
    EXPECT_GT(new_category_id, 0);
    
    // User can see the new category
    categories = Models::getAllCategories(*db);
    EXPECT_EQ(categories.size(), initial_count + 1);
    
    // User can use new category for expense
    int user_id = createSampleUser();
    int expense_id = Models::insertExpense(*db, "2025-10-25", new_category_id,
                                          "Test Expense", 50.0, user_id);
    EXPECT_GT(expense_id, 0);
    
    // Verify expense has correct category
    std::vector<Expense> expenses = Models::fetchExpensesByFilters(*db);
    ASSERT_EQ(expenses.size(), 1);
    EXPECT_EQ(expenses[0].category_name, "CustomCategory");
}

/**
 * USER STORY: As a user, I can manage users
 * Given: I want to track expenses for multiple people
 * When: I record expenses with different user names
 * Then: Each user's expenses are tracked separately
 */
TEST_F(ModelsTest, UserCanManageUsers) {
    // User records expenses for different people
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "Alice Lunch", 25.0, "Alice");
    ExpenseOperations::recordExpense(*db, "2025-10-26", "Food", "Bob Lunch", 30.0, "Bob");
    ExpenseOperations::recordExpense(*db, "2025-10-27", "Food", "Alice Dinner", 45.0, "Alice");
    
    // System tracks users separately
    std::vector<User> users = Models::getAllUsers(*db);
    EXPECT_EQ(users.size(), 2);
    
    // User can view expenses by specific user
    auto alice = Models::getUserByName(*db, "Alice");
    ASSERT_TRUE(alice.has_value());
    
    std::vector<Expense> alice_expenses = ExpenseOperations::viewExpensesByUser(*db, alice->id);
    EXPECT_EQ(alice_expenses.size(), 2);
    
    // Verify Alice's total
    ExpenseSummary alice_summary = ExpenseOperations::calculateSummary(*db, &alice_expenses);
    EXPECT_DOUBLE_EQ(alice_summary.total, 70.0);
}

/**
 * USER STORY: As a user, the system prevents invalid data
 * Given: I try to enter invalid expense data
 * When: I submit the data
 * Then: The system rejects it with an error
 * 
 * C++ FEATURE: Exception handling for validation
 */
TEST_F(ModelsTest, ExpenseValidationPreventsInvalidData) {
    // Invalid date format
    EXPECT_THROW(
        ExpenseOperations::recordExpense(*db, "invalid", "Food", "Test", 50.0, "User"),
        std::invalid_argument
    );
    
    // Negative amount
    EXPECT_THROW(
        ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "Test", -50.0, "User"),
        std::invalid_argument
    );
    
    // Zero amount
    EXPECT_THROW(
        ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "Test", 0.0, "User"),
        std::invalid_argument
    );
    
    // Empty title
    EXPECT_THROW(
        ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "", 50.0, "User"),
        std::invalid_argument
    );
    
    // Empty user name
    EXPECT_THROW(
        ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "Test", 50.0, ""),
        std::invalid_argument
    );
    
    // Verify no invalid expenses were created
    std::vector<Expense> expenses = ExpenseOperations::viewAllExpenses(*db);
    EXPECT_EQ(expenses.size(), 0);
}

/**
 * USER STORY: As a user, I can filter by amount range
 * Given: I have expenses of different amounts
 * When: I filter by amount range
 * Then: I see only expenses within that range
 */
TEST_F(ModelsTest, UserCanFilterByAmountRange) {
    // User records expenses with different amounts
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "Cheap", 10.0, "User");
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "Medium", 50.0, "User");
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "Expensive", 100.0, "User");
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "Very Expensive", 200.0, "User");
    
    // User filters by amount range ($40-$120)
    std::vector<Expense> filtered = ExpenseOperations::viewExpensesByAmount(*db, 40.0, 120.0);
    
    // User sees only expenses in range
    ASSERT_EQ(filtered.size(), 2);
    EXPECT_EQ(filtered[0].title, "Medium");
    EXPECT_EQ(filtered[1].title, "Expensive");
}

/**
 * USER STORY: As a user, I can filter by category
 * Given: I have expenses in different categories
 * When: I filter by specific categories
 * Then: I see only expenses in those categories
 */
TEST_F(ModelsTest, UserCanFilterByCategory) {
    // User records expenses in different categories
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "Lunch", 25.0, "User");
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Transport", "Taxi", 15.0, "User");
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Entertainment", "Movie", 20.0, "User");
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "Dinner", 40.0, "User");
    
    // User filters by Food category
    std::vector<std::string> categories = {"Food"};
    std::vector<Expense> filtered = ExpenseOperations::viewExpensesByCategory(*db, categories);
    
    // User sees only Food expenses
    ASSERT_EQ(filtered.size(), 2);
    EXPECT_EQ(filtered[0].category_name, "Food");
    EXPECT_EQ(filtered[1].category_name, "Food");
    
    // User filters by multiple categories
    categories = {"Transport", "Entertainment"};
    filtered = ExpenseOperations::viewExpensesByCategory(*db, categories);
    
    ASSERT_EQ(filtered.size(), 2);
}

/**
 * USER STORY: As a user, I can track expenses over time
 * Given: I record expenses over multiple days
 * When: I view my expense history
 * Then: I can see trends and patterns
 */
TEST_F(ModelsTest, UserCanTrackExpensesOverTime) {
    // User records expenses over a week
    ExpenseOperations::recordExpense(*db, "2025-10-20", "Food", "Monday Lunch", 15.0, "User");
    ExpenseOperations::recordExpense(*db, "2025-10-21", "Food", "Tuesday Lunch", 18.0, "User");
    ExpenseOperations::recordExpense(*db, "2025-10-22", "Food", "Wednesday Lunch", 20.0, "User");
    ExpenseOperations::recordExpense(*db, "2025-10-23", "Food", "Thursday Lunch", 17.0, "User");
    ExpenseOperations::recordExpense(*db, "2025-10-24", "Food", "Friday Lunch", 22.0, "User");
    
    // User views all expenses
    std::vector<Expense> all_expenses = ExpenseOperations::viewAllExpenses(*db);
    EXPECT_EQ(all_expenses.size(), 5);
    
    // User calculates weekly summary
    ExpenseSummary summary = ExpenseOperations::calculateSummary(*db, &all_expenses);
    EXPECT_DOUBLE_EQ(summary.total, 92.0);
    EXPECT_EQ(summary.count, 5);
    
    // User can view specific date range
    std::vector<Expense> midweek = ExpenseOperations::viewExpensesByDate(
        *db, "2025-10-22", "2025-10-23"
    );
    EXPECT_EQ(midweek.size(), 2);
    
    ExpenseSummary midweek_summary = ExpenseOperations::calculateSummary(*db, &midweek);
    EXPECT_DOUBLE_EQ(midweek_summary.total, 37.0);
}

/**
 * USER STORY: As a user, multiple users can track expenses independently
 * Given: Multiple people use the system
 * When: Each person records their expenses
 * Then: Each person's data is kept separate and can be viewed individually
 */
TEST_F(ModelsTest, MultipleUsersIndependentTracking) {
    // Alice records her expenses
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "Alice Breakfast", 10.0, "Alice");
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Transport", "Alice Bus", 5.0, "Alice");
    
    // Bob records his expenses
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "Bob Lunch", 20.0, "Bob");
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Entertainment", "Bob Movie", 15.0, "Bob");
    
    // Charlie records his expenses
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "Charlie Dinner", 30.0, "Charlie");
    
    // Each user can view their own expenses
    auto alice = Models::getUserByName(*db, "Alice");
    auto bob = Models::getUserByName(*db, "Bob");
    auto charlie = Models::getUserByName(*db, "Charlie");
    
    ASSERT_TRUE(alice.has_value());
    ASSERT_TRUE(bob.has_value());
    ASSERT_TRUE(charlie.has_value());
    
    std::vector<Expense> alice_expenses = ExpenseOperations::viewExpensesByUser(*db, alice->id);
    std::vector<Expense> bob_expenses = ExpenseOperations::viewExpensesByUser(*db, bob->id);
    std::vector<Expense> charlie_expenses = ExpenseOperations::viewExpensesByUser(*db, charlie->id);
    
    EXPECT_EQ(alice_expenses.size(), 2);
    EXPECT_EQ(bob_expenses.size(), 2);
    EXPECT_EQ(charlie_expenses.size(), 1);
    
    // Each user's totals are independent
    ExpenseSummary alice_summary = ExpenseOperations::calculateSummary(*db, &alice_expenses);
    ExpenseSummary bob_summary = ExpenseOperations::calculateSummary(*db, &bob_expenses);
    ExpenseSummary charlie_summary = ExpenseOperations::calculateSummary(*db, &charlie_expenses);
    
    EXPECT_DOUBLE_EQ(alice_summary.total, 15.0);
    EXPECT_DOUBLE_EQ(bob_summary.total, 35.0);
    EXPECT_DOUBLE_EQ(charlie_summary.total, 30.0);
    
    // Overall summary shows all users
    std::vector<Expense> all_expenses = ExpenseOperations::viewAllExpenses(*db);
    ExpenseSummary overall = ExpenseOperations::calculateSummary(*db, &all_expenses);
    
    EXPECT_DOUBLE_EQ(overall.total, 80.0);
    EXPECT_EQ(overall.count, 5);
    EXPECT_EQ(overall.by_user.size(), 3);
}

/**
 * USER STORY: As a user, I can see formatted output
 * Given: I have recorded expenses
 * When: I view expenses or summary
 * Then: The output is well-formatted and readable
 * 
 * C++ FEATURE: String formatting with Utils class
 */
TEST_F(ModelsTest, UserSeesFormattedOutput) {
    // User records expenses
    ExpenseOperations::recordExpense(*db, "2025-10-25", "Food", "Lunch", 25.50, "User");
    ExpenseOperations::recordExpense(*db, "2025-10-26", "Transport", "Taxi", 15.75, "User");
    
    // User views formatted expense list
    std::vector<Expense> expenses = ExpenseOperations::viewAllExpenses(*db);
    std::string expense_output = Utils::formatExpenseOutput(expenses);
    
    // Output contains formatted data
    EXPECT_NE(expense_output.find("$25.50"), std::string::npos);
    EXPECT_NE(expense_output.find("$15.75"), std::string::npos);
    EXPECT_NE(expense_output.find("Lunch"), std::string::npos);
    EXPECT_NE(expense_output.find("Taxi"), std::string::npos);
    
    // User views formatted summary
    ExpenseSummary summary = ExpenseOperations::calculateSummary(*db, &expenses);
    std::string summary_output = Utils::formatSummaryOutput(summary);
    
    // Summary contains formatted totals
    EXPECT_NE(summary_output.find("$41.25"), std::string::npos);
    EXPECT_NE(summary_output.find("EXPENSE SUMMARY"), std::string::npos);
    EXPECT_NE(summary_output.find("%"), std::string::npos);  // Percentages
}

/**
 * USER STORY: As a user, I can work with the system safely
 * Given: The system uses proper resource management
 * When: I perform operations
 * Then: Resources are properly cleaned up automatically
 * 
 * C++ FEATURE: RAII ensures automatic cleanup
 */
TEST(FunctionalTest, UserBenefitsFromRAII) {
    std::string test_path = "test_functional_raii.db";
    
    // User performs operations in a scope
    {
        Database db(test_path);
        db.initialize();
        
        // User records expenses
        ExpenseOperations::recordExpense(db, "2025-10-25", "Food", "Test", 50.0, "User");
        
        // Resources automatically cleaned up when scope ends
    }
    
    // User can reopen and continue working
    {
        Database db(test_path);
        std::vector<Expense> expenses = ExpenseOperations::viewAllExpenses(db);
        EXPECT_EQ(expenses.size(), 1);
    }
    
    // Cleanup
    std::remove(test_path.c_str());
}

